Fixes #19135 - Possibility to limit fields that are displayed

ofedoren opened the pull request:
Requires #275 to be merged first. PR already goes with that commit for testing.

The [issue](http://projects.theforeman.org/issues/19135) suggests --columns option while I suggest --only, because it works with all the adapters (not for table nor list commands only) and it's shorter :)

Also it accepts labels, which user sees, as arguments, so as mentioned in the issue e.g.:
`hammer subnet list --columns name,vlanid` will not work, while
`hammer --only name,"vlan id" subnet list` will work.

tstrachota commented on lib/hammer_cli/main.rb:
```diff
@@ -39,6 +39,9 @@ class MainCommand < AbstractCommand
     option ["--interactive"], "INTERACTIVE", _("Explicitly turn interactive mode on/off"),
       :format => HammerCLI::Options::Normalizers::Bool.new,
       :context_target => :interactive
+    option ["--only"], "FIELDS", _("Show selected columns or fields only"),
```
Placing the option in the main command isn't very intuitive. Could you move it to the base info and list commands? It also doesn't make sense to use it for create, update and delete.

tstrachota commented on lib/hammer_cli/output/output.rb:
```diff
@@ -69,5 +71,11 @@ def init_adapter(adapter_name)
+      return unless @context.key?(:only)
+      return if adapter.class <= HammerCLI::Output::Adapter::TreeStructure
+      labels = @context[:only].map(&:downcase)
+      definition.select_fields!(deep_select: true, labels: labels)
```
Using `select_fields!` will remove the fields from the command class for good. This will disallow using the option repeatedly in shell mode:
```
hammer shell
> host list --fields Id,Name
... will list the hosts
> host list --fields Id,Name,Subnet
... will break
```
Note that at the moment it's not possible to use the option in shell mode at all (see my previous comment).

tstrachota commented on lib/hammer_cli/output/output.rb:
```diff
@@ -69,5 +71,11 @@ def init_adapter(adapter_name)
 
+    def adjust_definition(definition)
+      return unless @context.key?(:only)
+      return if adapter.class <= HammerCLI::Output::Adapter::TreeStructure
```
Ideally we should deal with filtering options at one place only.

ofedoren commented:
updated.

Field sets are defined as new output definitions for commands. Field sets could be defined with `output(nil || output_definition, 'SET_NAME') do ... end`. Field set 'ALL' is a special one which is defined by default as one field set created from all currently available in `Command.field_sets`. All the fields currently defined with `output do ... end` go to the 'DEFAULT' field set.

for testing wth HammerCLIForeman:

Now it requires something like:
```
use_option :fields

output(nil, 'THIN') do
  field :id, _('Id')
  field :name, _('Name')
end
```

in commands such as `HammerCLIForeman::ListCommand` and `HammerCLIForeman::InfoCommand` to get the feature work. Also to enable 'thin' request we need to include
`      params_pruned['thin'] = true if current_output_definition.thin?`
in [request_params](https://github.com/theforeman/hammer-cli-foreman/blob/a6f940b32e77b863c678163e3c230c11737ef88e/lib/hammer_cli_foreman/commands.rb#L181)

I've prepared a PR with those changes for HammerCLIForeman. If there is a need, I can send it.

tstrachota commented:
Thank you @ofedoren. I'm a bit concerned about the design. Doesn't adding multiple output definitions go against your PR for output modification? https://github.com/theforeman/hammer-cli/pull/275

Wouldn't it be easier if each field would have notion about which field sets it should be printed in? It should then be just matter of filtering the fields before they're printed and still only one output definition would be used.

ofedoren commented:
@tstrachota, you're right, it was unnecessary complicated, so I redesignied according to your suggestions:
 - Field sets are specified in options: `field :key, :label, :type, sets: ['SETNAME']`
 - Only one output definition is used
 - Fields filter does the whole work

[Updated docs](https://github.com/theforeman/hammer-cli/blob/1884b36c6f1bfdb3b894de4b623e9d8d28238849/doc/creating_commands.md)

Still requires `params_pruned['thin'] = true if output.adapter.thin_context?`
in [request_params](https://github.com/theforeman/hammer-cli-foreman/blob/a6f940b32e77b863c678163e3c230c11737ef88e/lib/hammer_cli_foreman/commands.rb#L181) for thin request and `use_option :fields` in `ListCommand`s and `InfoCommand`s 

Also, I'm afraid we need to specify `'THIN'` set explicitly for `Id` and `Name` fields. I was thinking about adding `Id` and `Name` fields to `THIN` set by default from core just looking if field name is `id` or `name`, but I don't think it's a good idea.

tstrachota commented:
This needs a rebase.

ofedoren commented:
@tstrachota, rebased.

mbacovsky commented on doc/creating_commands.md:
````diff
@@ -100,11 +100,14 @@ Option declaration is the same as in clamp so please read it's
 ```ruby
 class HelloCommand < HammerCLI::AbstractCommand
 
+  use_option :option_name, :or_names
````
@ofedoren Here I'd expect naming conflict with the option as both could be called `--name`, also from the example output it is not apparent what it does. There is just one --name option in the output. Maybe it would be better not to mention `use_option` here and create separate subsection  `####Predefined options` next to `#### Deprecated options`. May make sense to explain how it works, how to use it and list available predefined options. With `:fields` we can add brief description with reference to `Printing hash records` section for more details.

mbacovsky commented on lib/hammer_cli/options/option_definition.rb:
```diff
@@ -27,9 +27,9 @@ class OptionDefinition < Clamp::Option::Definition
-        self.value_formatter = options.delete(:format) || HammerCLI::Options::Normalizers::Default.new
-        self.context_target = options.delete(:context_target)
-        self.deprecated_switches = options.delete(:deprecated)
+        self.value_formatter = options[:format] || HammerCLI::Options::Normalizers::Default.new
```
Is there a reason to remove the cleanup of the options?

ofedoren replied:
If I leave the cleanup then the options hash for the `option` in predefined options will be empty, which makes the predefined option unusable for other commands.

mbacovsky commented on lib/hammer_cli/output/adapter/abstract.rb:
```diff
@@ -41,10 +42,19 @@ def print_collection(fields, collection)
     end
 
+    def thin_context?
+      return false if @context[:fields].nil?
```
Wouldn't `@context[:fields] ==['THIN']` do the same as the two lines here?

mbacovsky commented on lib/hammer_cli/output/adapter/tree_structure.rb:
```diff
@@ -34,15 +34,11 @@ def prepare_message(msg, msg_params = {})
     def render_fields(fields, data)
-      fields = field_filter.filter(fields)
-      fields = displayable_fields(fields, data)
+      fields = filter_fields(fields).filter_by_classes
```
Calling the filters repeats in all the adapters, could it be DRY'ed out a bit?

mbacovsky commented on lib/hammer_cli/output/output.rb:
```diff
@@ -25,13 +25,15 @@ def print_error(msg, details=nil, msg_params = {}, options = {})
 
     def print_record(definition, record)
       adapter.print_record(definition.fields, record) if appropriate_verbosity?(:record)
+      adapter.reset_context
```
Why it is necessary to remove the :fields from context?

ofedoren replied:
I do it to be able to reset fields selection in case of using hammer in shell mode.

mbacovsky commented:
@ofedoren, well done!. I like how it works and that the logic is kept quite separated in the adapter. I've left a few comments inline. A few more here:
- maybe it is time to uncover hammer-cli-foreman part, feel free to mark it [WIP] if not ready
- regarding THIN and request_params - did you tried to handle it in an option source? It could be possible to avoid changes in `request_params`
- regarding THIN and marking params explicitly - did you consider using SEARCHABLES to distinguish the right options? I'm not sure how the thin fields are defined in API and if they match to what we have in the searchables. Also we should have a plan for hammer-cli-katello
- currently it is not visible for a user what fields are in which sets. I'd love to see some help extension when --fields is defined on the command showing e.g. table like:
```
Options:
  --fileds    FIELDS  Show specified fileds or predefined filed sets only. (See below)
 
...

Predefined field sets:
+--------+-----+---------+------+
| Fields | ALL | DEFAULT | THIN |
+--------+-----+---------+------+
| Id     |  x  |    x    |  x   |
| Name   |  x  |    x    |  x   |
| Param1 |  x  |    x    |      |
| Param2 |  x  |         |      |
+--------+-----+---------+------+
```

ofedoren commented:
@mbacovsky, updated. See inline comments, also:

[Updated docs](https://github.com/theforeman/hammer-cli/blob/ce55a3afc949477263a5b25f8f8369f744dc60a2/doc/creating_commands.md#predefined-options)

As you suggested, I use now option sources instead of changing `request_params` directly, see https://github.com/theforeman/hammer-cli-foreman/pull/407

But still there is a problem: I'm afraid we need to specify `'THIN'` set explicitly for `Id` and `Name` fields in all the command output definitions (in plugins as well).

mbacovsky commented:
@ofedoren I'm okay with adding the help extension in separate PR. Otherwise it looks good now.
For the THIN issue - I'd really like to avoid setting it manually. I have an idea:
in the command we have `self.output` method for output definition. We could extend it in foreman commands with something like:
```ruby
output_definition.update_field_set('THIN', searchables.for(resource) + ['id])
```
`update_field_set` would iterate over the fields in definition and update the set on the matching fields.
Would it make sense?

ofedoren commented:
@mbacovsky, updated. i've added the `update_field_set` method as you suggested ([docs](https://github.com/theforeman/hammer-cli/blob/3ccd557a588da40de9839f828e3c12848ec67681/doc/commands_modification.md)).

The usage of it see: https://github.com/theforeman/hammer-cli-foreman/pull/407

mbacovsky commented on lib/hammer_cli/output/field_filter.rb:
```diff
@@ -1,21 +1,87 @@
+      return true if labels.include?(label)
+
+      labels.each do |l|
+        return true if l.start_with?("#{label}/") || label.start_with?(l)
```
@ofedoren finally I was able to figure out how this is supposed to work and fixed the condition to autocomplete with asterisk `l.start_with?("#{label}/") || label.match(%r{^#{l.gsub(/\*/, '.*')}(|\/.*)$})`. I simplified it a bit also and have
```ruby
   def include_by_label?(labels, label)
      labels.any? do |l|
        l.start_with?("#{label}/") || label.match(%r{^#{l.gsub(/\*/, '.*')}(|\/.*)$})
      end
    end
```
Now it works with:
- `host info --id 4 --fields 'id,name,network'`
- `host info --id 4 --fields 'id,name,network*'`
- `host info --id 4 --fields 'id,name,network/domain'`

What do you think?

tstrachota commented:
Congratulations @ofedoren :beers:! It was very close to celebrate a one-year anniversary of this PR! :birthday: :hammer: :slightly_smiling_face:
